src/nvme/mi.c

```diff
@@ -419,6 +418,80 @@ int nvme_mi_admin_xfer(nvme_mi_ctrl_t ctrl,
 	return 0;
 }
 
+int nvme_mi_admin_admin_passthru(nvme_mi_ctrl_t ctrl, __u8 opcode, __u8 flags,
+				 __u16 rsvd, __u32 nsid, __u32 cdw2, __u32 cdw3,
+				 __u32 cdw10, __u32 cdw11, __u32 cdw12,
+				 __u32 cdw13, __u32 cdw14, __u32 cdw15,
+				 __u32 data_len, void *data, __u32 metadata_len,
+				 void *metadata, __u32 timeout_ms, __u32 *result)
+{
+	/* Input parameters flags, rsvd, metadata, metadata_len are not used */
+	struct nvme_mi_admin_resp_hdr resp_hdr;
+	struct nvme_mi_admin_req_hdr req_hdr;
+	struct nvme_mi_resp resp;
+	struct nvme_mi_req req;
+	int rc;
+	if(data_len > 4096) {
+		nvme_msg(ctrl->ep->root, LOG_ERR, "nvme_mi_admin_admin_passthru"
+			"does not supports data_len over 4096 bytes.\n");
+		errno = EINVAL;
+		return -1;
+	}
+
+	if(timeout_ms > nvme_mi_ep_get_timeout(ctrl->ep)) {
+		/* Set timeout if user needs a bigger timeout */
+		nvme_mi_ep_set_timeout(ctrl->ep, timeout_ms);
+	}
+
+	nvme_mi_admin_init_req(&req, &req_hdr, ctrl->id, opcode);
+	req_hdr.cdw1 = cpu_to_le32(nsid);
+	req_hdr.cdw2 = cpu_to_le32(cdw2);
+	req_hdr.cdw3 = cpu_to_le32(cdw3);
+	req_hdr.cdw10 = cpu_to_le32(cdw10);
+	req_hdr.cdw11 = cpu_to_le32(cdw11);
+	req_hdr.cdw12 = cpu_to_le32(cdw12);
+	req_hdr.cdw13 = cpu_to_le32(cdw13);
+	req_hdr.cdw14 = cpu_to_le32(cdw14);
+	req_hdr.cdw15 = cpu_to_le32(cdw15);
+	req_hdr.doff = 0;
+	if (data_len != 0) {
+		req_hdr.dlen = cpu_to_le32(data_len);
+		/* Bit 0 set to 1 means DLEN contains a value */
+		req_hdr.flags = 0x1;
+	}
+
+	/* write data */
+	if((opcode & 0x3) == NVME_DATA_TFR_HOST_TO_CTRL) {
```

Comment:
`NVME_DATA_TFR_HOST_TO_CTRL`  looks really like the wrong enum type. This enum is for the `nvme_get_feature_length2` function.

And why not just introduce static inline function instead the original macro?

Reply:
The `enum nvme_data_tfr` is defined according to the description: opcode[1:0]"Indicates the data transfer direction of the command. All options to the command shall transfer data as specified or transfer no data. All commands, including vendor specific commands, shall follow this convention: 00b = no data transfer; 01b = host to controller; 10b = controller to host; 11b = bidirectional." in NVMe spec. So this enum apply to all Admin commands, not only set_feature/get_feature.

The `static inline function` looks like:
```
static inline int nvme_get_admin_data_direction(__u8 opcode) {
    return opcode & 0x3;
}
```

The use case looks like:
```
if(nvme_get_admin_data_direction(opcode) == NVME_DATA_TFR_HOST_TO_CTRL) {
```

VS current sytle:
```
if((opcode & 0x3) == NVME_DATA_TFR_HOST_TO_CTRL) {
```

I prefer my current style as it will save me some typing effort.

Reply:
Typing effort is almost never a good optimisation; this code will be read *many* more times than it is typed. Please optimise for clarity instead.

Reply:
> The enum nvme_data_tfr is defined according to the description:

Yes and no. It's an argument type for the `nvme_get_feature_length2`. That is also the reason why it lives in `utils.h`. But the biggest clue is that we don't define any of the values, we let the compiler choose. So you would first need to update `enum nvme_data_tfr` accordingly. 

I asked to transform `IS_DIRECTION_WRITE` to a function. The resulting code would be something
like `if (nvme_is_direction_write(op)) { ...}` or whatever the function is supposed be named to match
the current coding style.

Reply:
1. Defined values in `enum nvme_data_tfr`. No sure if you need to move it from `util.h` to other header file.
2. Changed usage to:
```
       int direction = opcode & 0x3;
       bool has_write_data = (direction == NVME_DATA_TFR_HOST_TO_CTRL)
               && (data != NULL) && (data_len != 0);
       bool has_read_data = (direction == NVME_DATA_TFR_CTRL_TO_HOST)
               && (data != NULL) && (data_len != 0);
       if (direction == NVME_DATA_TFR_BIDIRECTIONAL) {...
```
Please let me know if you still want to add 3 helper functions(`nvme_is_direction_write(op)`, `nvme_is_direction_read(op)`, `nvme_is_bidirection(op)`)

Reply:
This is difficult to read. Fixed it myself.